deepblast/viterbi_cuda.py

```diff
@@ -0,0 +1,239 @@
+import torch
+import numba
+from numba import cuda
+from math import exp, log
+import torch.nn as nn
+from deepblast.constants import m_, x_, y_, s_
+
+# FIXME: Refactor cuda utility and variables
+torch.autograd.set_detect_anomaly(True)
+
+max_cols = 2048
+float_type = numba.float32
+max_S = 8
+tpb = 32  # threads per block
+
+
+@cuda.jit(device=True)
+def _soft_max_device(X, S, A):
+    M = X[0]
+    for i in range(1, S):
+        M = max(M, X[i])
+    sumA = 0
+    for i in range(S):
+        sumA += exp(X[i] - M)
+    M += log(sumA)
+    for i in range(S):
+        A[i] = exp(X[i] - M)
+    return M
+
+
+@cuda.jit(device=True)
+def _forward_pass_device(theta, A, pos, Q):
+    maxargs = cuda.local.array(max_S, float_type)
+    V = cuda.local.array((2, max_cols, max_S), float_type)
+    N, M, S = theta.shape
+
+    neg_inf = -1e8
+    last, curr = 0, 1
+    for j in range(M + 1):
+        for k in range(S):
+            V[last, j, k] = neg_inf
+            V[curr, j, k] = neg_inf
+
+    for k in range(S):
+        V[last, 0][k] = 0.0
+
+    for i in range(1, N + 1):
+        for j in range(1, M + 1):
+            for k in range(S):
+                di = pos[k][0]
+                dj = pos[k][1]
+
+                for L in range(S):
+                    maxargs[L] = V[curr + di, j + dj, L] + \
+                                 A[i - 1, j - 1, k, L]
+
+                V[curr, j, k] = _soft_max_device(maxargs, S, Q[i, j, k])
+
+            for k in range(0, S):
+                V[curr, j, k] += theta[i - 1, j - 1, k]
+
+        last = curr
+        curr = 1 - curr
+
+    Vt = _soft_max_device(V[last, M], S, Q[N + 1, M + 1, 0])
+    return Vt
```

Comment:
@blackwer , just realized that this function also needs to return the entire matrix `V`. I'm not entirely sure what exactly max_cols and max_S are -- is it easy to assemble the full matrix from this?